pkg/ruler/storage/wal/wal.go

```diff
@@ -677,6 +689,11 @@ func (a *appender) Commit() error {
 		buf = buf[:0]
 	}
 
+	// Notify so that reader waiting for it can read without needing to wait for next read ticker.
+	if a.writeNotified() != nil {
+		a.writeNotified().Notify()
+	}
```

Comment:
This is a little clunky; how about just passing a `Notify()` function to the appender? The appender doesn't need any knowledge of the `wlog.WriteNotified`.

Reply:
I wouldn't have thought about this fix. This is cool. That said I'm a little confused. Shouldn't we implement `Notify()` and pass it on?

Reply:
Agree on the clunkiness :) 

It comes from 

```
type appender struct {
	w             *Storage
	writeNotified func() wlog.WriteNotified
	series        []record.RefSeries
	samples       []record.RefSample
	exemplars     []record.RefExemplar
}
```

`writeNotified` being `func()` type.

It got bit tricky because we initialize few things during `NewStorage()` and few other things during `initialize()` call. 

wal is created during `wal.NewStorage` but the remoteStorage (that has `Notify()`) only created during `initialize`. We have to it as func type because we can only know if `wal.Storage` has write notify set that way. Also note, `wal.Storage` uses `appender` as a pool.

We can make it more clean by refactoring few things on `NewStorage` and `initialize`. But AFAIK not worth it now?

Reply:
Also about the nit: `SetWriteNotified` -> `SetWriteNotifier`. 

I think they calling it `WriteNotified` and not `WriteNotifier` because, `WriteNotifier` is confusing in a sense, is it notifying the "write" that is done? or notifying the "writer".

We want to notify the "write" that is actually done (potentially to readers waiting for it).